File: ai_dj/llm.py
"""LLM client with two providers.

- "go":  opencode Zen Go subscription (https://opencode.ai/zen/go/v1). Needs
         OPENCODE_API_KEY. mimo-v2.6-flash can hear audio via input_audio parts.
- "local": a llama.cpp server (default http://127.0.0.1:8080/v1). No key. Audio
         support depends on the loaded model; if the model can't take audio the
         call falls back to text (current script as context).

Shared by the `probe` prototype and the `live` loop.
"""

import base64
import json
import os
import re
import subprocess

ZEN_GO_BASE = "https://opencode.ai/zen/go/v1"
LOCAL_BASE_DEFAULT = os.environ.get("LLAMA_BASE_URL", "http://127.0.0.1:8080/v1")
DEFAULT_MODEL = "mimo-v2.6-flash"

_REFERENCE_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                               "reference", "sonic_pi.md")


def _load_reference():
    try:
        with open(_REFERENCE_PATH) as f:
            return f.read()
    except OSError:
        return ""


_DJ_INSTRUCTIONS = """You are the live-coding engine of a generative radio DJ.
You hear a short audio sample of the current music and evolve the set.

You are given: the CURRENT Sonic Pi script (what is playing), a short AUDIO
SAMPLE of its output, and optionally USER FEEDBACK from the listener.

Your job: return an UPDATED version of the script that evolves the music
SMOOTHLY. Respond with ONLY valid JSON, no markdown:

{
  "hearing": {
    "tempo_bpm": 0,
    "key": "",
    "chords": [],
    "energy": "",
    "mood": "",
    "structure": ""
  },
  "decision": {
    "action": "",
    "why": "",
    "change_to": ""
  },
  "ruby": ""
}

HARD RULES:
- Smooth transitions. NEVER jump key, tempo or mood. Keep the current key,
  tempo and feel unless the user feedback explicitly asks otherwise, and then
  move GRADUALLY (change one element at a time, not the whole piece).
- Preserve structure. Keep the existing live_loops; change at most 1-2 elements
  per update (add/remove a layer, adjust a filter, vary a pattern, nudge energy).
- Return the COMPLETE, runnable script (all live_loops), not a diff.
- Every musical choice gets a preceding "# why" comment, referencing what you
  heard or the feedback you acted on.
- Only Sonic Pi DSL. No explanation outside the JSON.

Use the reference below to pick real synths, samples, FX and patterns — make
musically deliberate choices, not random ones.

# Sonic Pi Reference
"""

_DJ_INSTRUCTIONS_LEAN = """You are the live-coding engine of a generative radio DJ.
You hear a short audio sample of the current music and evolve the set smoothly.
Given the CURRENT script, an AUDIO SAMPLE, and optional USER FEEDBACK, return an
UPDATED runnable script as JSON:

{"hearing": {"tempo_bpm": 0, "key": "", "chords": [], "energy": "", "mood": "", "structure": ""},
 "decision": {"action": "", "why": "", "change_to": ""},
 "ruby": ""}

Rules: keep key/tempo/mood unless feedback says otherwise; change at most 1-2
elements per update; return the COMPLETE script (all live_loops); precede each
musical choice with a "# why" comment; valid Sonic Pi DSL only; JSON only.
"""


def _system(reference=True):
    if reference:
        return _DJ_INSTRUCTIONS + _load_reference()
    return _DJ_INSTRUCTIONS_LEAN


SYSTEM = _system(True)


def load_env_key(env_path):
    if env_path and os.path.exists(env_path):
        for line in open(env_path):
            m = re.match(r"export\s+OPENCODE_API_KEY\s*=\s*(.+)", line)
            if m:
                return m.group(1).strip().strip('"').strip("'")
    return os.environ.get("OPENCODE_API_KEY", "")


def _audio_part(path):
    if path.endswith(".wav"):
        down = path.rsplit(".", 1)[0] + "_16k.mp3"
        subprocess.run(
            ["ffmpeg", "-y", "-i", path, "-ar", "16000", "-ac", "1",
             "-codec:a", "libmp3lame", "-qscale:a", "7", down],
            capture_output=True, check=True)
        path = down
    b64 = base64.b64encode(open(path, "rb").read()).decode()
    return {"type": "input_audio", "input_audio": {"data": b64, "format": "mp3"}}


# -- provider plumbing ------------------------------------------------------

def _base_url(provider, base_url=None):
    if base_url:
        return base_url
    return LOCAL_BASE_DEFAULT if provider == "local" else ZEN_GO_BASE


def _post_chat(provider, base_url, api_key, body, session_id=None, timeout=300):
    url = _base_url(provider, base_url).rstrip("/") + "/chat/completions"
    headers = ["-H", "Content-Type: application/json"]
    if api_key:
        headers += ["-H", f"Authorization: Bearer {api_key}"]
    if session_id:
        headers += ["-H", f"x-opencode-session: {session_id}"]
    p = subprocess.run(
        ["curl", "-s", "-X", "POST", url, *headers, "-d", "@-"],
        input=json.dumps(body), capture_output=True, text=True, timeout=timeout)
    if p.returncode != 0:
        raise RuntimeError(f"curl failed: {p.stderr.strip()[:300]}")
    if not p.stdout.strip():
        raise RuntimeError(f"empty response from {url}")
    data = json.loads(p.stdout)
    if "error" in data:
        raise RuntimeError(f"API error: {data['error']}")
    return data


def list_local_models(base_url=None, timeout=10):
    """Return the model ids served by a llama.cpp (or any OpenAI-compatible)
    server, or [] if unreachable."""
    url = _base_url("local", base_url).rstrip("/") + "/models"
    try:
        p = subprocess.run(["curl", "-s", "--max-time", str(timeout), url],
                           capture_output=True, text=True, timeout=timeout + 5)
    except Exception:
        return []
    if p.returncode != 0 or not p.stdout.strip():
        return []
    try:
        data = json.loads(p.stdout)
    except ValueError:
        return []
    return [m.get("id") for m in data.get("data", []) if m.get("id")]


# -- public calls -----------------------------------------------------------

def _parse(data, model):
    choice = data["choices"][0]
    msg = choice["message"]
    content = msg.get("content")
    if not content:
        reason = choice.get("finish_reason")
        raise RuntimeError(
            f"model returned no content (finish={reason}); "
            f"try raising max_tokens or a smaller reference")
    parsed = _extract_json(content)
    parsed["_usage"] = data.get("usage", {})
    parsed["_model"] = model
    return parsed


def listen_and_decide_prompt(prompt, key, model=DEFAULT_MODEL, session_id=None,
                             provider="go", base_url=None, reference=True):
    """Generate an initial script from a vibe prompt (no audio input)."""
    body = {
        "model": model,
        "messages": [
            {"role": "system", "content": _system(reference)},
            {"role": "user", "content": (
                f"Create the FIRST Sonic Pi script for a live set with this vibe: "
                f'"{prompt}". Fill "hearing" with what this vibe implies '
                f'(tempo, key, energy, mood), "decision" with the starting plan, '
                f'and put the full playable script in "ruby".')},
        ],
        "max_tokens": 16000,
        "temperature": 0.8,
    }
    sid = session_id or f"ai-dj-{os.getpid()}"
    data = _post_chat(provider, base_url, key, body, session_id=sid)
    return _parse(data, model)


def listen_and_decide(audio_path, key, model=DEFAULT_MODEL, session_id=None,
                      prompt=None, provider="go", base_url=None,
                      current_script=None, feedback=None, reference=True):
    """Send the current script + audio sample (+ feedback), get the evolution.

    Context is bounded per call: system (cached) + current script + audio +
    feedback. No chat history is carried, so a session can run for hours.
    """
    intro = []
    if current_script:
        intro.append("Here is the CURRENT Sonic Pi script (what is playing now):\n"
                     f"```ruby\n{current_script}\n```")
    intro.append("Here is a 10-second sample of its current output. Evolve the "
                 "music smoothly.")
    if feedback:
        intro.append(f"USER FEEDBACK for this iteration (act on it): {feedback}")
    user_content = [{"type": "text", "text": "\n\n".join(intro)},
                    _audio_part(audio_path)]

    guide = f'\nSTYLE GUIDE (keep the whole set in this vibe): "{prompt}"' if prompt else ""
    body = {
        "model": model,
        "messages": [
            {"role": "system", "content": _system(reference) + guide},
            {"role": "user", "content": user_content},
        ],
        "max_tokens": 16000,
        "temperature": 0.8,
    }
    sid = session_id or f"ai-dj-{os.getpid()}"
    try:
        data = _post_chat(provider, base_url, key, body, session_id=sid)
        return _parse(data, model)
    except Exception as e:
        if provider != "local" or current_script is None:
            raise
        # Local model may not accept audio: retry text-only with the script.
        print(f"[llm] local audio failed ({e}); retrying text-only")
        body["messages"][1]["content"] = (
            f"Your model cannot hear audio. Evolve this CURRENT script musically, "
            f"smoothly.\n\n```ruby\n{current_script}\n```")
        data = _post_chat(provider, base_url, key, body, session_id=sid)
        return _parse(data, model)


def _extract_json(text):
    decoder = json.JSONDecoder()
    idx = text.find("{")
    if idx < 0:
        raise ValueError(f"no JSON in model reply: {text[:300]}")
    obj, _ = decoder.raw_decode(text[idx:])
    return obj

File: packages/dj/src/llm.ts
/** LLM client with two providers.
 *
 * - "go":  opencode Zen subscription (https://opencode.ai/zen/go/v1). Needs
 *          OPENCODE_API_KEY. The model hears audio via input_audio parts.
 * - "local": a llama.cpp server (default http://127.0.0.1:8080/v1). No key.
 *
 * Frugality rules baked in here (they are the subscription budget):
 * - The system prompt is byte-identical every call (instructions + reference),
 *   so it is fully prompt-cached. Everything variable goes in the user message.
 * - The evolve call asks for ONE layer patch, not the whole script.
 *
 * Ported from ai_dj/llm.py. The server has no audio of its own, so audio is an
 * optional base64 part instead of an ffmpeg pipeline.
 */

import { existsSync, readFileSync } from "node:fs";
import { asMarkdown } from "./palette";

export const ZEN_GO_BASE = "https://opencode.ai/zen/go/v1";
export const LOCAL_BASE_DEFAULT =
  process.env.LLAMA_BASE_URL ?? "http://127.0.0.1:8080/v1";
export const DEFAULT_MODEL = "mimo-v2.6-flash";
export const DEFAULT_LANG = "sonic_pi";
export const LANGS = ["sonic_pi", "strudel"] as const;
export type Lang = (typeof LANGS)[number];

const EVOLVE: Record<string, string> = {
  sonic_pi: `You are the live-coding engine of a generative radio DJ.
You evolve ONE layer of a Sonic Pi live set at a time, smoothly and gradually.

Given: the session state, the layer to change, the direction to take, that
layer's current code (if any), a 10s audio sample, and optional user feedback.

Return ONLY valid JSON, no markdown:
{
  "hearing": {"tempo_bpm": 0, "key": "", "chords": [], "energy": "", "mood": "", "structure": ""},
  "decision": {"action": "", "why": ""},
  "op": {"kind": "add|modify|remove", "layer": "<the given layer>"},
  "ruby": "<ONE live_loop for that layer, with # why comments>"
}

HARD RULES:
- Change ONLY the given layer. Never touch other layers.
- Do NOT set use_bpm; the session tempo is fixed.
- Keep the session key and mode; stay in the current vibe.
- This is one incremental step, not a new song.
- "ruby" must be a single valid Sonic Pi live_loop.
- Precede each musical choice with a "# why" comment.
- No text outside the JSON.

# Sonic Pi Reference
`,
  strudel: `You are the live-coding engine of a generative radio DJ.
You evolve ONE layer of a Strudel live set at a time, smoothly and gradually.

Given: the session state, the layer to change, the direction to take, that
layer's current pattern (if any), a 10s audio sample, and optional user feedback.

Return ONLY valid JSON, no markdown:
{
  "hearing": {"tempo_bpm": 0, "key": "", "chords": [], "energy": "", "mood": "", "structure": ""},
  "decision": {"action": "", "why": ""},
  "op": {"kind": "add|modify|remove", "layer": "<the given layer>"},
  "ruby": "<ONE Strudel pattern expression for that layer>"
}

HARD RULES:
- Change ONLY the given layer. Never touch other layers.
- Do NOT call setcpm/setcps; the session tempo is fixed.
- Keep the session key and mode; stay in the current vibe.
- This is one incremental step, not a new song.
- "ruby" must be a single Strudel pattern expression: no stack(), no $: labels.
- Choose timbres only from the instrument palette.
- Precede each musical choice with a "// why" comment.
- No text outside the JSON.

# Strudel Reference
`,
};

const SEED: Record<string, string> = {
  sonic_pi: `You are the live-coding engine of a generative radio DJ.
Create the FIRST script for a new live set, based on the vibe given.

Return ONLY valid JSON, no markdown:
{
  "hearing": {"tempo_bpm": 0, "key": "", "chords": [], "energy": "", "mood": "", "structure": ""},
  "decision": {"action": "", "why": ""},
  "ruby": "<the complete playable script>"
}

HARD RULES:
- Set one use_bpm at the top and keep it for the whole set.
- Build one live_loop per element (kick, bass, hats, pad, lead, FX) and sync them.
- Every musical choice gets a preceding "# why" comment.
- Valid Sonic Pi DSL only. No text outside the JSON.

# Sonic Pi Reference
`,
  strudel: `You are the live-coding engine of a generative radio DJ.
Create the FIRST set for a new live session, based on the vibe given.

Return ONLY valid JSON, no markdown:
{
  "hearing": {"tempo_bpm": 0, "key": "", "chords": [], "energy": "", "mood": "", "structure": ""},
  "decision": {"action": "", "why": ""},
  "ruby": "<the complete playable Strudel script>"
}

HARD RULES:
- Start with setcpm(<bpm>/4) and keep that tempo for the whole set.
- Build the set as ONE expression: stack(...) with one pattern per layer
  (kick, bass, hats, pad, lead, FX).
- Choose timbres only from the instrument palette.
- Every musical choice gets a preceding "// why" comment.
- Valid Strudel only. No $: labels. No text outside the JSON.

# Strudel Reference
`,
};

// Variant used when the session already has an identity (an archetype, or the
// set being replaced): the model composes inside that tempo/key/mode.
const SEED_ANCHORED: Record<string, string> = {
  sonic_pi: `You are the live-coding engine of a generative radio DJ.
Write the music for an EXISTING live set, based on the vibe given.

The session already has a fixed tempo, key and mode. Compose inside them.

Return ONLY valid JSON, no markdown:
{
  "hearing": {"tempo_bpm": 0, "key": "", "chords": [], "energy": "", "mood": "", "structure": ""},
  "decision": {"action": "", "why": ""},
  "ruby": "<the complete playable script>"
}

HARD RULES:
- Do NOT set use_bpm; keep the session's tempo.
- Keep the session's key and mode; stay in the current vibe.
- Build one live_loop per element (kick, bass, hats, pad, lead, FX) and sync them.
- Every musical choice gets a preceding "# why" comment.
- Valid Sonic Pi DSL only. No text outside the JSON.

# Sonic Pi Reference
`,
  strudel: `You are the live-coding engine of a generative radio DJ.
Write the music for an EXISTING set, based on the vibe given.

The session already has a fixed tempo, key and mode. Compose inside them.

Return ONLY valid JSON, no markdown:
{
  "hearing": {"tempo_bpm": 0, "key": "", "chords": [], "energy": "", "mood": "", "structure": ""},
  "decision": {"action": "", "why": ""},
  "ruby": "<the complete playable Strudel script>"
}

HARD RULES:
- Do NOT call setcpm; keep the session's tempo.
- Keep the session's key and mode; stay in the current vibe.
- Build the set as ONE expression: stack(...) with one pattern per layer
  (kick, bass, hats, pad, lead, FX).
- Choose timbres only from the instrument palette.
- Every musical choice gets a preceding "// why" comment.
- Valid Strudel only. No $: labels. No text outside the JSON.

# Strudel Reference
`,
};

export interface SystemPromptOptions {
  referenceText?: string;
  anchored?: boolean;
}

export function systemPrompt(
  kind: "evolve" | "seed",
  lang: string = DEFAULT_LANG,
  opts: SystemPromptOptions = {},
): string {
  const prompts = opts.anchored ? SEED_ANCHORED : kind === "evolve" ? EVOLVE : SEED;
  const base = prompts[lang] ?? prompts[DEFAULT_LANG];
  if (opts.referenceText == null) return base; // without the (large) reference
  let text = base + opts.referenceText;
  if (lang === "strudel") {
    // always append the current palette so the model cannot generate
    // instrument ids that render silence
    text += "\n\n# Current instrument palette\n" + asMarkdown();
  }
  return text;
}

export function loadEnvKey(envPath?: string | null): string {
  if (envPath) {
    try {
      if (existsSync(envPath)) {
        for (const line of readFileSync(envPath, "utf8").split("\n")) {
          const m = line.match(/export\s+OPENCODE_API_KEY\s*=\s*(.+)/);
          if (m) return m[1].trim().replace(/^["']|["']$/g, "");
        }
      }
    } catch { /* fall through to the environment */ }
  }
  return process.env.OPENCODE_API_KEY ?? "";
}

/** Tolerate leading prose; the model often chats before the JSON. Like the
 * decoder the Python client uses, parse one object and ignore trailing text.
 */
export function extractJson(text: string): Record<string, unknown> {
  const idx = text.indexOf("{");
  if (idx < 0) throw new Error(`no JSON in model reply: ${text.slice(0, 300)}`);
  const src = text.slice(idx);
  let depth = 0;
  let inString = false;
  let escaped = false;
  for (let i = 0; i < src.length; i++) {
    const ch = src[i];
    if (inString) {
      if (escaped) escaped = false;
      else if (ch === "\\") escaped = true;
      else if (ch === '"') inString = false;
    } else if (ch === '"') {
      inString = true;
    } else if (ch === "{") {
      depth++;
    } else if (ch === "}") {
      depth--;
      if (depth === 0) return JSON.parse(src.slice(0, i + 1)) as Record<string, unknown>;
    }
  }
  throw new Error(`no complete JSON object in model reply: ${text.slice(0, 300)}`);
}

export type FetchFn = (
  url: string,
  init: { method: string; headers: Record<string, string>; body: string },
) => Promise<{ ok: boolean; json(): Promise<unknown> }>;

export interface CallOptions {
  provider?: string;
  baseUrl?: string;
  reasoning?: string;
  sessionId?: string;
  fetchFn?: FetchFn;
}

function resolveBaseUrl(provider: string, baseUrl?: string): string {
  if (baseUrl) return baseUrl;
  return provider === "local" ? LOCAL_BASE_DEFAULT : ZEN_GO_BASE;
}

async function postChat(
  provider: string,
  baseUrl: string | undefined,
  apiKey: string,
  body: Record<string, unknown>,
  opts: { sessionId?: string; fetchFn?: FetchFn } = {},
): Promise<Record<string, unknown>> {
  const url = `${resolveBaseUrl(provider, baseUrl).replace(/\/$/, "")}/chat/completions`;
  const headers: Record<string, string> = { "Content-Type": "application/json" };
  if (apiKey) headers["Authorization"] = `Bearer ${apiKey}`;
  const sid = opts.sessionId ?? `ai-dj-${process.pid}`;
  if (sid) headers["x-opencode-session"] = sid;
  const run = opts.fetchFn ?? fetch;
  const res = await run(url, { method: "POST", headers, body: JSON.stringify(body) });
  const data = (await res.json()) as Record<string, unknown>;
  if ("error" in data) throw new Error(`API error: ${JSON.stringify(data["error"])}`);
  return data;
}

function parseCompletion(data: Record<string, unknown>, model: string): Record<string, unknown> {
  const choice = (data["choices"] as Array<Record<string, unknown>>)[0];
  const msg = choice["message"] as Record<string, unknown>;
  const content = msg["content"];
  if (!content || typeof content !== "string") {
    throw new Error(`model returned no content (finish=${choice["finish_reason"]})`);
  }
  const parsed = extractJson(content) as Record<string, unknown>;
  parsed["_usage"] = (data["usage"] as unknown) ?? {};
  parsed["_model"] = model;
  return parsed;
}

export interface EvolveOptions extends CallOptions {
  key: string;
  model?: string;
  stateContext?: string;
  layer?: string;
  direction?: string;
  layerCode?: string | null;
  feedback?: string | null;
  referenceText?: string;
  lang?: string;
  wholeSet?: boolean;
  audioBase64?: string | null;
}

export async function evolveLayer(opts: EvolveOptions): Promise<Record<string, unknown>> {
  const {
    key, model = DEFAULT_MODEL, stateContext = "", layer = "", direction = "",
    layerCode = null, feedback = null, provider = "go", baseUrl,
    referenceText, reasoning = "none", lang = DEFAULT_LANG, wholeSet = false,
    audioBase64 = null,
  } = opts;

  let system: string;
  let lines: string[];
  if (wholeSet) {
    system = systemPrompt("seed", lang, { referenceText, anchored: true });
    lines = [
      `Session state: ${stateContext}`,
      `Evolve the whole set — direction: ${direction}.`,
      "The set is ONE script. Change it in that direction and return the "
      + "complete updated script, keeping everything you are not changing "
      + "(including any samples()/setcps() lines and the overall structure).",
      `Current script:\n\`\`\`\n${layerCode ?? ""}\n\`\`\``,
    ];
    if (feedback) lines.push(`User feedback (act on it): ${feedback}`);
    lines.push("Here is a 10s sample of the current output.");
  } else {
    system = systemPrompt("evolve", lang, { referenceText });
    lines = [
      `Session state: ${stateContext}`,
      `Change layer: ${layer}`,
      `Direction: ${direction}`,
    ];
    lines.push(layerCode
      ? `Current code for '${layer}':\n\`\`\`ruby\n${layerCode}\n\`\`\``
      : `Current code for '${layer}': none yet — create it.`);
    if (feedback) lines.push(`User feedback (act on it): ${feedback}`);
    lines.push("Here is a 10s sample of the current output.");
  }

  const userContent: unknown = audioBase64
    ? [
      { type: "text", text: lines.join("\n\n") },
      { type: "input_audio", input_audio: { data: audioBase64, format: "mp3" } },
    ]
    : lines.join("\n\n");

  const body = {
    model,
    messages: [
      { role: "system", content: system },
      { role: "user", content: userContent },
    ],
    max_tokens: 8000,
    temperature: 0.8,
    reasoning_effort: reasoning,
  };
  const sid = opts.sessionId ?? `ai-dj-${process.pid}`;
  try {
    const data = await postChat(provider, baseUrl, key, body,
      { sessionId: sid, fetchFn: opts.fetchFn });
    return parseCompletion(data, model);
  } catch (e) {
    if (provider !== "local" || !layerCode) throw e;
    const fallback = {
      ...body,
      messages: [
        body.messages[0],
        {
          role: "user",
          content:
            `Your model cannot hear audio. Evolve layer '${layer}' ` +
            `(${direction}) of this set.\n\n\`\`\`ruby\n${layerCode}\n\`\`\``,
        },
      ],
    };
    const data = await postChat(provider, baseUrl, key, fallback,
      { sessionId: sid, fetchFn: opts.fetchFn });
    return parseCompletion(data, model);
  }
}

export interface SeedOptions extends CallOptions {
  prompt: string;
  key: string;
  model?: string;
  referenceText?: string;
  lang?: string;
  session?: { bpm: number; key: string; mode: string } | null;
}

/** Write the music for a vibe prompt.
 *
 * session: an optional {bpm, key, mode} identity to compose inside (an
 * archetype the session was cast from). Without it the model chooses freely.
 */
export async function seedScript(opts: SeedOptions): Promise<Record<string, unknown>> {
  const {
    prompt, key, model = DEFAULT_MODEL, sessionId, provider = "go", baseUrl,
    referenceText, reasoning = "none", lang = DEFAULT_LANG, session = null,
  } = opts;
  const anchored = Boolean(session);
  const system = systemPrompt("seed", lang, { referenceText, anchored });
  let user = `Write the music for a live set with this vibe: "${prompt}".`;
  if (anchored && session) {
    user += `\n\nSession identity (compose inside it): ${session.bpm}bpm, ` +
      `key ${session.key}, ${session.mode}.`;
  }
  const body = {
    model,
    messages: [
      { role: "system", content: system },
      { role: "user", content: user },
    ],
    max_tokens: 16000,
    temperature: 0.8,
    reasoning_effort: reasoning,
  };
  const data = await postChat(provider, baseUrl, key, body,
    { sessionId, fetchFn: opts.fetchFn });
  return parseCompletion(data, model);
}
